File: forms/vaccinations-checklist/full-stack-with-loco-tera-htmx-alpine/tests/engine/vaccination_grader_test.rs
use vaccinations_checklist_tera_crate::engine::types::*;
use vaccinations_checklist_tera_crate::engine::vaccination_grader::{grade, run_rules};
use vaccinations_checklist_tera_crate::engine::vaccination_rules::all_rules;

/// A baseline fully-compliant healthcare worker case.
fn fully_compliant_healthcare_worker() -> AssessmentData {
    AssessmentData {
        demographics: Demographics {
            first_name: "Jane".to_string(),
            last_name: "Doe".to_string(),
            date_of_birth: "1985-03-12".to_string(),
            sex: "female".to_string(),
            weight: Some(70.0),
            height: Some(170.0),
            bmi: Some(24.2),
            occupation: "Nurse".to_string(),
            occupation_category: "healthcare".to_string(),
            employer: "NHS Trust".to_string(),
        },
        childhood_immunisations: ChildhoodImmunisations {
            mmr_dose1: "yes".to_string(),
            mmr_dose2: "yes".to_string(),
            dtp_primary_course: "yes".to_string(),
            dtp_booster: "yes".to_string(),
            polio_primary_course: "yes".to_string(),
            polio_booster: "yes".to_string(),
            ..ChildhoodImmunisations::default()
        },
        occupational_vaccines: OccupationalVaccines {
            hepatitis_b_course: "yes".to_string(),
            hepatitis_b_anti_body_level: "adequate".to_string(),
            varicella_vaccine: "yes".to_string(),
            varicella_history: "yes".to_string(),
            bcg_vaccine: "yes".to_string(),
            ..OccupationalVaccines::default()
        },
        covid19_vaccination: Covid19Vaccination {
            covid_primary_course: "yes".to_string(),
            covid_booster1: "yes".to_string(),
            ..Covid19Vaccination::default()
        },
        influenza_vaccination: InfluenzaVaccination {
            flu_vaccine_current_season: "yes".to_string(),
            ..InfluenzaVaccination::default()
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_assessment_is_not_fully_immunised() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_ne!(result.compliance_status, "fully-immunised");
}

#[test]
fn fully_compliant_healthcare_worker_is_fully_immunised() {
    let data = fully_compliant_healthcare_worker();
    let result = grade(&data);
    assert_eq!(result.compliance_status, "fully-immunised");
    assert_eq!(result.overall_risk, "low");
    assert!(result.childhood_complete);
    assert!(result.occupational_complete);
    assert!(result.covid_complete);
    assert!(result.flu_current);
}

#[test]
fn anaphylaxis_history_fires_ctx_002_critical() {
    let mut data = fully_compliant_healthcare_worker();
    data.vaccination_history.adverse_reaction_severity = "anaphylaxis".to_string();
    let fired = run_rules(&data);
    let ctx002 = fired.iter().find(|r| r.id == "CTX-002").expect("CTX-002 should fire");
    assert_eq!(ctx002.grade, 4);
}

#[test]
fn healthcare_worker_missing_hep_b_is_non_compliant_high_risk() {
    let mut data = fully_compliant_healthcare_worker();
    data.occupational_vaccines.hepatitis_b_course = "no".to_string();
    let result = grade(&data);
    assert_eq!(result.compliance_status, "non-compliant");
    assert_eq!(result.overall_risk, "high");
    assert!(result.fired_rules.iter().any(|r| r.id == "OCC-001"));
}

#[test]
fn active_exposure_incident_yields_critical_risk() {
    let mut data = fully_compliant_healthcare_worker();
    data.schedule_compliance.active_exposure_incident = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.overall_risk, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "CMP-001"));
}

#[test]
fn live_vaccine_contraindicated_is_contraindicated() {
    let mut data = fully_compliant_healthcare_worker();
    data.contraindications_allergies.live_vaccine_contraindicated = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.compliance_status, "contraindicated");
}

#[test]
fn travel_planned_without_yellow_fever_fires_trv_001() {
    let mut data = fully_compliant_healthcare_worker();
    data.travel_vaccines.travel_planned = "yes".to_string();
    data.travel_vaccines.yellow_fever_vaccine = "no".to_string();
    let fired = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "TRV-001"));
}

#[test]
fn measles_igg_negative_fires_ser_001() {
    let mut data = fully_compliant_healthcare_worker();
    data.serology_immunity_testing.measles_ig_g = "negative".to_string();
    let fired = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "SER-001"));
}

#[test]
fn missing_childhood_mmr_appears_in_missing_vaccinations() {
    let mut data = fully_compliant_healthcare_worker();
    data.childhood_immunisations.mmr_dose1 = "no".to_string();
    let result = grade(&data);
    assert!(result
        .missing_vaccinations
        .iter()
        .any(|v| v == "MMR dose 1"));
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

File: forms/vaccinations-checklist/front-end-with-svelte/src/lib/engine/vaccination-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateVaccinationGrade } from './vaccination-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { vaccinationRules } from './vaccination-rules';
import type { AssessmentData } from './types';

/**
 * A blank checklist literal. Kept local to the test so the engine suite has no
 * dependency on the Svelte store (which imports `$app/environment`).
 */
function blank(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', weight: null, height: null, bmi: null, occupation: '', occupationCategory: '', employer: '' },
		vaccinationHistory: { hasVaccinationRecord: '', recordSource: '', recordSourceOther: '', previousAdverseReaction: '', adverseReactionDetails: '', adverseReactionVaccine: '', adverseReactionSeverity: '', immunocompromised: '', immunocompromisedDetails: '', pregnantOrPlanning: '' },
		childhoodImmunisations: { mmrDose1: '', mmrDose1Date: '', mmrDose2: '', mmrDose2Date: '', dtpPrimaryCourse: '', dtpPrimaryDate: '', dtpBooster: '', dtpBoosterDate: '', polioPrimaryCourse: '', polioPrimaryDate: '', polioBooster: '', polioBoosterDate: '', hibVaccine: '', hibVaccineDate: '', menCVaccine: '', menCVaccineDate: '', menACWYVaccine: '', menACWYVaccineDate: '', pcvVaccine: '', pcvVaccineDate: '', notes: '' },
		occupationalVaccines: { hepatitisBCourse: '', hepatitisBCourseDate: '', hepatitisBDosesReceived: null, hepatitisBAntiBodyLevel: '', bcgVaccine: '', bcgVaccineDate: '', bcgScarPresent: '', varicellaVaccine: '', varicellaVaccineDate: '', varicellaHistory: '', hepatitisAVaccine: '', hepatitisAVaccineDate: '', typhoidVaccine: '', typhoidVaccineDate: '', rabiesVaccine: '', rabiesVaccineDate: '', notes: '' },
		travelVaccines: { travelPlanned: '', travelDestination: '', travelDepartureDate: '', travelReturnDate: '', yellowFeverVaccine: '', yellowFeverVaccineDate: '', yellowFeverCertificate: '', japaneseEncephalitisVaccine: '', japaneseEncephalitisDate: '', tickBorneEncephalitisVaccine: '', tickBorneEncephalitisDate: '', choleraVaccine: '', choleraVaccineDate: '', meningococcalACWYTravel: '', meningococcalACWYTravelDate: '', malariaProphylaxis: '', malariaProphylaxisDrug: '', notes: '' },
		covid19Vaccination: { covidPrimaryCourse: '', covidPrimaryVaccineType: '', covidDose1Date: '', covidDose2Date: '', covidBooster1: '', covidBooster1Date: '', covidBooster1Type: '', covidBooster2: '', covidBooster2Date: '', covidBooster2Type: '', covidAutumnBooster: '', covidAutumnBoosterDate: '', totalCovidDoses: null, covidAdverseReaction: '', covidAdverseReactionDetails: '', notes: '' },
		influenzaVaccination: { fluVaccineCurrentSeason: '', fluVaccineCurrentDate: '', fluVaccineType: '', fluVaccinePreviousSeason: '', fluVaccineAnnualRecipient: '', fluHighRiskGroup: '', fluHighRiskReason: '', fluAdverseReaction: '', fluAdverseReactionDetails: '', notes: '' },
		contraindicationsAllergies: { eggAllergy: '', eggAllergySeverity: '', gelatinAllergy: '', neomycinAllergy: '', latexAllergy: '', yeastAllergy: '', pegPolysorbateAllergy: '', otherVaccineAllergies: '', historyOfGBS: '', gbsDetails: '', onImmunosuppressants: '', immunosuppressantDetails: '', onBloodProductsRecent: '', bloodProductsDetails: '', liveVaccineContraindicated: '', liveVaccineContraindicationReason: '', notes: '' },
		serologyImmunityTesting: { hepBSurfaceAntibody: '', hepBSurfaceAntibodyLevel: null, hepBSurfaceAntibodyDate: '', varicellaIgG: '', varicellaIgGDate: '', measlesIgG: '', measlesIgGDate: '', rubellaIgG: '', rubellaIgGDate: '', mumpsIgG: '', mumpsIgGDate: '', hepAIgG: '', hepAIgGDate: '', tetanusAntibody: '', tetanusAntibodyDate: '', tbIGRAResult: '', tbIGRADate: '', mantouxResult: '', mantouxIndurationMm: null, notes: '' },
		scheduleCompliance: { complianceStatus: '', vaccinesDue: '', vaccinesOverdue: '', catchUpPlanRequired: '', catchUpPlanDetails: '', nextVaccinationDate: '', nextVaccinationType: '', occupationalHealthClearance: '', occupationalHealthClearanceDate: '', exposureRiskLevel: '', activeExposureIncident: '', activeExposureDetails: '', consentForVaccination: '', consentDate: '', notes: '' }
	};
}

/** A fully immunised, low-risk patient. */
function fullyImmunised(): AssessmentData {
	const d = blank();
	d.demographics.lastName = 'Test';
	d.demographics.dateOfBirth = '1985-01-01';
	d.childhoodImmunisations.mmrDose1 = 'yes';
	d.childhoodImmunisations.mmrDose2 = 'yes';
	d.childhoodImmunisations.dtpPrimaryCourse = 'yes';
	d.childhoodImmunisations.dtpBooster = 'yes';
	d.childhoodImmunisations.polioPrimaryCourse = 'yes';
	d.childhoodImmunisations.polioBooster = 'yes';
	d.covid19Vaccination.covidPrimaryCourse = 'yes';
	d.covid19Vaccination.covidBooster1 = 'yes';
	d.influenzaVaccination.fluVaccineCurrentSeason = 'yes';
	return d;
}

describe('calculateVaccinationGrade', () => {
	it('classifies a complete schedule as fully immunised, low risk', () => {
		const r = calculateVaccinationGrade(fullyImmunised());
		expect(r.complianceStatus).toBe('fully-immunised');
		expect(r.overallRisk).toBe('low');
		expect(r.childhoodComplete).toBe(true);
		expect(r.covidComplete).toBe(true);
		expect(r.fluCurrent).toBe(true);
		expect(r.firedRules.length).toBe(0);
	});

	it('flags childhood gaps as partially immunised', () => {
		const d = fullyImmunised();
		d.childhoodImmunisations.mmrDose1 = 'no';
		const r = calculateVaccinationGrade(d);
		expect(r.childhoodComplete).toBe(false);
		expect(r.complianceStatus).toBe('partially-immunised');
		expect(r.firedRules.some((x) => x.id === 'CH-001')).toBe(true);
	});

	it('marks a healthcare worker with no COVID primary course as non-compliant, critical risk', () => {
		const d = fullyImmunised();
		d.demographics.occupationCategory = 'healthcare';
		d.occupationalVaccines.hepatitisBCourse = 'yes';
		d.occupationalVaccines.hepatitisBAntiBodyLevel = 'adequate';
		d.occupationalVaccines.varicellaHistory = 'yes';
		d.covid19Vaccination.covidPrimaryCourse = 'no';
		const r = calculateVaccinationGrade(d);
		expect(r.firedRules.some((x) => x.id === 'COV-002' && x.grade === 4)).toBe(true);
		expect(r.complianceStatus).toBe('non-compliant');
		// A grade-4 rule escalates the overall risk to critical.
		expect(r.overallRisk).toBe('critical');
	});

	it('classifies a non-healthcare flu gap as partially immunised, moderate risk', () => {
		const d = fullyImmunised();
		d.influenzaVaccination.fluVaccineCurrentSeason = 'no';
		const r = calculateVaccinationGrade(d);
		expect(r.firedRules.some((x) => x.id === 'FLU-001' && x.grade === 2)).toBe(true);
		expect(r.complianceStatus).toBe('partially-immunised');
		expect(r.overallRisk).toBe('moderate');
	});

	it('treats live-vaccine contraindication as contraindicated', () => {
		const d = fullyImmunised();
		d.contraindicationsAllergies.liveVaccineContraindicated = 'yes';
		const r = calculateVaccinationGrade(d);
		expect(r.complianceStatus).toBe('contraindicated');
	});

	it('escalates an active exposure incident to critical risk', () => {
		const d = fullyImmunised();
		d.scheduleCompliance.activeExposureIncident = 'yes';
		const r = calculateVaccinationGrade(d);
		expect(r.overallRisk).toBe('critical');
	});
});

describe('detectAdditionalFlags', () => {
	it('raises a high-priority flag for vaccine anaphylaxis history', () => {
		const d = fullyImmunised();
		d.vaccinationHistory.adverseReactionSeverity = 'anaphylaxis';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-ANAPH-001' && f.priority === 'high')).toBe(true);
	});

	it('returns no flags for a clean record', () => {
		expect(detectAdditionalFlags(fullyImmunised()).length).toBe(0);
	});
});

describe('vaccinationRules', () => {
	it('every rule has a 1-4 grade and an evaluator', () => {
		for (const rule of vaccinationRules) {
			expect(rule.grade).toBeGreaterThanOrEqual(1);
			expect(rule.grade).toBeLessThanOrEqual(4);
			expect(typeof rule.evaluate).toBe('function');
		}
	});
});
